mmdet3d/ops/paconv/src/assign_score_withk_cuda.cu

```diff
@@ -0,0 +1,214 @@
+                                                    const long* knn_idx,
+                                                    float* output) {
+
+    // ----- parallel loop for B, N1, K and O ---------
+    long i = blockIdx.x * blockDim.x + threadIdx.x;
+    if (i >= B*N1*K*O) return;
+    // ------- loop for M ----------
+    for (int m = 0; m < M; m++) {
+        int b = (int)(i / (O * N1 * K));
+        int o = (int)(i % (O * N1 * K) / (N1 * K));
+        int n = (int)(i % (N1 * K) / K);
+        int k = (int)(i % K);
+        int cn = (int) knn_idx[b*K*N1 + n*K + 0]; //The first neighbor is the center point
+        int kn = (int) knn_idx[b*K*N1 + n*K + k];
+        if (kn >= N0 || kn < 0) { // if index overflows, it is out of the neighborhood range
+            continue;
+        }
+        assert (b < B);
+        assert (kn < N0);
+        assert (cn < N0);
+        assert (o < O);
+        assert (n < N1);
+        atomicAdd(output + b*N1*O*K + o*N1*K + n*K + k,
+            points[b*N0*M*O + kn*M*O + m*O + o] * scores[b*N1*K*M + n*K*M + k*M + m]
+                - centers[b*N0*M*O + cn*M*O + m*O + o] * scores[b*N1*K*M + n*K*M + k*M + m]);
+    }
+}
+
+
+__global__ void assign_score_withk_backward_points_kernel(const int B, const int N0, const int N, const int M,
+                                                            const int K, const int O, const int aggregate,
+                                                            const float* grad_out,
+                                                            const float* scores,
+                                                            const long* knn_idx,
+                                                            float* grad_points,
+                                                            float* grad_centers) {
+
+    // ----- parallel loop for B, M, O ---------
+    long i = blockIdx.x * blockDim.x + threadIdx.x;
+    if (i >= B*M*O) return;
+    int b = (int)(i / (M * O));
+    int m = (int)(i % (M * O) / O);
+    int o = (int)(i % O);
+
+    // ----- loop for N,K ---------
+    for (int n = 0; n < N; n++) {
+        for (int k = 0; k < K; k++) {
+            int kn = knn_idx[b*N*K + n*K + k];
+            int cn = knn_idx[b*N*K + n*K + 0];
+            if (kn >= N0 || kn < 0) { // if index overflows, it is out of the neighborhood range
+                continue;
+            }
+            atomicAdd(grad_points + b*N0*M*O + kn*M*O + m*O + o,
+                scores[b*N*K*M + n*K*M + k*M + m] * grad_out[b*O*N*K + o*N*K + n*K + k]);
+            atomicAdd(grad_centers + b*N0*M*O + cn*M*O + m*O + o,
+                - scores[b*N*K*M + n*K*M + k*M + m] * grad_out[b*O*N*K + o*N*K + n*K + k]);
+            //grad_points[b*N*M*O + kn*M*O + m*O + o] += 2 * scores[b*N*K*M + n*K*M + k*M + m] * grad_out[b*O*N + o*N + n];
+            //grad_points[b*N*M*O + n*M*O + m*O + o] -= scores[b*N*K*M + n*K*M + k*M + m] * grad_out[b*O*N + o*N + n];
```

Comment:
Clean these unused comments? Out of curiosity, the 2 in the first equation corresponds to sth. in the line 104-105 or just a typo?

Reply:
I think they are just typos.